pkg/abcupdater/version_test.go

```diff
@@ -206,3 +207,44 @@ To disable notifications for this new version, set SAMPLE_APP_1_IGNORE_VERSIONS=
 		})
 	}
 }
+
+// Note: These tests rely on timing and could be flaky if breakpoints are used.
+func Test_asyncFunctionCall(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name  string
+		input func() (string, error)
+		want  string
+	}{
+		{
+			name:  "happy_path",
+			input: func() (string, error) { return "done", nil },
+			want:  "done\n",
+		},
+		{
+			name:  "error_path_still_returns",
+			input: func() (string, error) { return "", fmt.Errorf("failed") },
+			want:  "",
+		},
+		{
+			// This test will take a long time due to forcing timeout.
+			name:  "timeout_gets_applied",
+			input: func() (string, error) { time.Sleep(9999 * time.Hour); return "should_not_execute", nil },
+			want:  "",
+		},
+		// TODO: would like a clean way to test the canceled context case, not sure
+		// how to make it not leak into rest of test cases
+	}
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			outBuf := bytes.Buffer{}
+			resultFunc := asyncFunctionCall(context.Background(), tc.input, &outBuf)
+			resultFunc()
```

Comment:
It might be worth verifying that nothing is written to outBuf until resultFunc() is called. That's one of the important properties of asyncFunctionCall.

You might have to wait for 5 milliseconds or something just to give the background goroutine time to print (which it shouldn't), like:

```go
			outBuf := bytes.Buffer{}
			resultFunc := asyncFunctionCall(context.Background(), tc.input, &outBuf)
			time.Sleep(5 * time.Millisecond)
			resultFunc()
```

Technically, in the failure case, this read would be a data race (since the writer and reader aren't synchronized). It's probably OK since there's only a data race if `asyncFunctionCall` is writing to outBuf before it's supposed to.

Reply:
Added a test, I'm not a fan of the race in the test, but not sure if there is a good way around it.